### PR TITLE
Fix Weighted panic

### PR DESCRIPTION
**Pull request by khasanovbi**
Please take a look.

Take is divided into 2 steps index search and after that Reweight with 0.
1. I added checks for heap over bounds during tree traversal.
2. I removed last and left because in each iteration of the loop we always make a move to either the left or right subtree and this branch never reached.
3. Instead of two implementations of reweight in the `Take` and `Reweight` functions, I left only one.
4. Instead of calling reset - O(n) when you check a negative weight, I do a weight calculation as in reset, but through sift up only the desired nodes for O(log(n)). It also removed the `scalar.EqualWithinAbsOrRel` check, since we are now summing up zeros instead of accumulating a float error.

https://github.com/gonum/gonum/issues/1866

### CODE REVIEW

**Comment by kortschak** [comment]
I'll take a look on the weekend. From an initial look, I'd like more detail on the fix in the description and less code movement.

**Comment by kortschak** [comment]
> can I change the golang version for staticcheck to 1.18?

Where do you want to do this?

**Comment by khasanovbi** [comment]
> Where do you want to do this?

[Here](https://github.com/gonum/gonum/blob/master/.github/workflows/staticcheck.yml#L16). I want to set it to v1 to use v1.3.0

File: .github/workflows/staticcheck.yml [hunk 1]
```diff
@@ -13,6 +13,6 @@ jobs:
     - uses: actions/checkout@v2
       with:
         fetch-depth: 1
-    - uses: dominikh/staticcheck-action@v1.2.0
+    - uses: dominikh/staticcheck-action@v1
```
Comment (kortschak) [inline .github/workflows/staticcheck.yml:16]: This should be in a separate PR.

Comment (khasanovbi) [reply .github/workflows/staticcheck.yml]: Moved this commit to https://github.com/gonum/gonum/pull/1868

Comment (kortschak) [reply .github/workflows/staticcheck.yml]: Thanks

File: stat/sampleuv/weighted.go [hunk 2]
```diff
@@ -54,70 +52,89 @@ func NewWeighted(w []float64, src rand.Source) Weighted {
 // already taken.
 func (s Weighted) Len() int { return len(s.weights) }
 
+func (s Weighted) findIndex(r float64) int {
```
Comment (kortschak) [inline stat/sampleuv/weighted.go:55]: Conventionally we place caller before callee, but in this case I'd prefer that this block be put back in `Take`; it's not called from more than one location and doesn't significantly complexify the calling function.

So, as
```
	r *= s.heap[0]
	i := 0
	for {
		r -= s.weights[i]
		if r < 0 {
			break // Fall within item i.
		}
...
```
after `r` decl in `Take`.

File: stat/sampleuv/weighted.go [hunk 3]
```diff
@@ -54,70 +52,89 @@ func NewWeighted(w []float64, src rand.Source) Weighted {
...
+
+	for {
+		r -= s.weights[i]
+		if r < 0 {
+			break // Fall within item i.
+		}
+
+		ni := i<<1 + 1 // Move to left child.
+		// Left node should exist, because r is non-negative, but there are could be floating point errors,
+		// so we check index explicitly.
```
Comment (kortschak) [inline stat/sampleuv/weighted.go:66]:
```suggestion
		// Left node should exist, because r is non-negative,
		// but there could be floating point errors, so we
		// check index explicitly.
```
(narrow comments are preferred within code)

File: stat/sampleuv/weighted.go [hunk 4]
```diff
@@ -54,70 +52,89 @@ func NewWeighted(w []float64, src rand.Source) Weighted {
...
+		if ni >= len(s.heap) {
+			break
+		}
+
+		i = ni
+
+		d := s.heap[i]
+		if r >= d {
+			// If enough r to pass left child try to move to the right child.
```
Comment (kortschak) [inline stat/sampleuv/weighted.go:75]:
```suggestion
			// If there is enough r to pass left child try to
			// move to the right child.
```

File: stat/sampleuv/weighted.go [hunk 5]
```diff
@@ -54,70 +52,89 @@ func NewWeighted(w []float64, src rand.Source) Weighted {
...
-	idx++
-	for idx > 0 {
-		s.heap[idx-1] -= w
-		idx >>= 1
+	s.weights[idx] = w
+
+	for {
+		w = s.weights[idx]
+		// We want to keep heap state here like after reset call.
+		// Therefore, we sum weights in such order, because floating point sum not associative.
```
Comment (kortschak) [inline stat/sampleuv/weighted.go:119]: Move this to above (attached to) the for loop as
```
	// We want to keep the heap state here consistent
	// with the result of a reset call. So we sum
	// weights in the same order, since floating point
	// addition is not associative.
```

File: stat/sampleuv/weighted_test.go [hunk 6]
```diff
@@ -266,3 +266,48 @@ func TestWeightedNoResample(t *testing.T) {
 		}
 	}
 }
+
+func calculateNonZeroWeights(weights []float64) int {
```
Comment (kortschak) [inline stat/sampleuv/weighted_test.go:270]: Please move this to below `FuzzWeightedTake`.

File: stat/sampleuv/weighted_test.go [hunk 7]
```diff
@@ -266,3 +266,48 @@ func TestWeightedNoResample(t *testing.T) {
 		}
 	}
 }
+
```
Comment (kortschak) [inline stat/sampleuv/weighted_test.go:269]: Please add a test for the cases in the issue
```
var issue1866Tests = []struct {
	weights []float64
	seed    uint64
}{
	{weights: []float64{72000, 166.66666666666666}, seed: 202},
	{weights: []float64{0.6666666666666666, 20412, 0}, seed: 0},
}

// See https://github.com/gonum/gonum/issues/1866
func TestIssue1866(t *testing.T) {
	for i, test := range issue1866Tests {
		w := NewWeighted(test.weights, rand.NewSource(test.seed))
		for j := 0; j < len(test.weights)+1; j++ {
			if panicked(func() { w.Take() }) {
				t.Errorf("unexpected panic for test %d iteration %d", i, j)
			}
		}
	}
}

func panicked(fn func()) (yes bool) {
	defer func() {
		if recover() != nil {
			yes = true
		}
	}()
	fn()
	return false
}
```

File: stat/sampleuv/weighted_test.go [hunk 8]
```diff
@@ -266,3 +266,48 @@ func TestWeightedNoResample(t *testing.T) {
...
+	f.Fuzz(func(t *testing.T, w1 float64, w2 float64) {
+		if w1 < 0 || w2 < 0 {
+			t.Skip()
+		}
+
+		weights := []float64{w1, w2}
+
+		weighted := NewWeighted(weights, rand.NewSource(0))
+
+		expectedTakenSize := calculateNonZeroWeights(weights)
```
Comment (kortschak) [inline stat/sampleuv/weighted_test.go:293]:
```suggestion
		got := calculateNonZeroWeights(weights)
```

File: stat/sampleuv/weighted_test.go [hunk 9]
```diff
@@ -266,3 +266,48 @@ func TestWeightedNoResample(t *testing.T) {
...
+		}
+
+		if expectedTakenSize != len(takenSet) {
+			t.Errorf(
+				"unexpected taken set size: expected=%d, actual=%d, taken=%v",
+				expectedTakenSize,
+				len(takenSet),
+				takenSet,
+			)
+		}
```
Comment (kortschak) [inline stat/sampleuv/weighted_test.go:311]: I don't think we need to keep a set, we expect n samples from an n-sized weighted, so let's use that. 

```suggestion
		want := calculateNonZeroWeights(weights)
		var got int
		for {
			_, ok := weighted.Take()
			if !ok {
				break
			}
			got++
		}

		if got != want {
			t.Errorf("unexpected taken set size: got=%d, want=%d", got, want)
		}
```

File: stat/sampleuv/weighted_test.go [hunk 10]
```diff
@@ -266,3 +266,48 @@ func TestWeightedNoResample(t *testing.T) {
...
+	for _, w := range weights {
+		if w != 0 {
+			count++
+		}
+	}
+
+	return count
+}
+
+func FuzzWeighted_Take(f *testing.F) {
```
Comment (kortschak) [inline stat/sampleuv/weighted_test.go:281]:
```suggestion
func FuzzWeightedTake(f *testing.F) {
```

**Review by kortschak — COMMENTED** [review]
Thanks. Mainly style issues.

File: stat/sampleuv/weighted_test.go (hunk 7 above)
Comment (khasanovbi) [reply stat/sampleuv/weighted_test.go:269]: Fixed, but maybe it would be better to prefer subtests with t.Run?

File: stat/sampleuv/weighted_test.go (hunk 10 above)
Comment (khasanovbi) [reply stat/sampleuv/weighted_test.go]: Fixed

File: stat/sampleuv/weighted_test.go (hunk 9 above)
Comment (khasanovbi) [reply stat/sampleuv/weighted_test.go]: Fixed

File: stat/sampleuv/weighted_test.go (hunk 6 above)
Comment (khasanovbi) [reply stat/sampleuv/weighted_test.go]: Fixed

File: stat/sampleuv/weighted.go (hunk 5 above)
Comment (khasanovbi) [reply stat/sampleuv/weighted.go]: Fixed

File: stat/sampleuv/weighted.go (hunk 4 above)
Comment (khasanovbi) [reply stat/sampleuv/weighted.go]: Fixed

File: stat/sampleuv/weighted.go (hunk 3 above)
Comment (khasanovbi) [reply stat/sampleuv/weighted.go]: Fixed

File: stat/sampleuv/weighted.go (hunk 2 above)
Comment (khasanovbi) [reply stat/sampleuv/weighted.go]: Fixed

File: stat/sampleuv/weighted.go [hunk 11]
```diff
@@ -6,8 +6,6 @@ package sampleuv
 
 import (
 	"golang.org/x/exp/rand"
-
-	"gonum.org/v1/gonum/floats/scalar"
 )
```
Comment (kortschak) [inline stat/sampleuv/weighted.go:9]: Since this is now a single import
```
import "golang.org/x/exp/rand"
```

File: stat/sampleuv/weighted_test.go (hunk 7 above)
Comment (kortschak) [reply stat/sampleuv/weighted_test.go:269]: This will do; having the test table in the func body makes it more cluttered and the tests are simple enough not to justify sub-tests.

Please move this addition to above the fuzz test.

**Review by kortschak — COMMENTED** [review]
LGTM after minor nits resolved. Thanks

**Review by kortschak — APPROVED** [review]
Thanks
